### DISCUSSION TITLE
Advise on feasibility of application integration (dynamic keys/actions?)

### DISCUSSION THREAD

**Discussion by celsochang** [elgatosf/elgato#91] [Help] [answered]
Hello,

I have a Java application which I would like to communicate with StreamDeck. 

My use case is:
1. The User selects a specific number of users from a list to participate on an activity - each user having a number assigned
2. Each user would be assigned a key on StreamDeck. And each key would be displaying that user's number
3. Each time the User presses the key, the application would receive the event with the user's number and process it
4. Later at the start of a new activity the User repeats steps 1-3, and the list of keys on StreamDeck is refreshed. The list of numbers can grow or shrink according to the users participating on the next activity.

Is it possible to update the keys without using the Elgato StreamDeck app? 

I've looked through the SDK documentation and looked at some YouTube videos and still can't figure out if what I want to do is possible or not. Could you please advise if it's feasible and would be great if you could tell me (high-level) how to achieve it or point me to some useful resources?

Thank you.

**Comment by nekename** [comment]
Not sure about Java, but for Python, Node, the browser, and Rust, there are third-party libraries available for direct access and control of Stream Deck devices.

**Comment by ChekTek** [comment, marked as answer]
This is definitely possible with a Stream Deck plugin, but that will require using the Stream Deck application. Our first-party SDK does not support controlling the hardware without the Stream Deck application at this time.

For a plugin, you can use a single action to populate a full Stream Deck profile, which you can bundle in your plugin.

https://docs.elgato.com/sdk/plugins/manifest#profiles

This will allow you to control all the keys at once. You can use the `willAppear` event to cache the context (or ID) of the keys on your profile, which you can reference later in your refresh step. I would also use the `willDisappear` event to clear your cached values. 

https://docs.elgato.com/sdk/plugins/events-received#willappear

Then for the property inspector you'll likely want to use global settings to keep track of things, so that you could make them available to all the keys of the profile.
